File: app/src/main/java/com/rmambulancier/nova/NovaNotificationListenerService.java
package com.rmambulancier.nova;

import android.app.Notification;
import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.media.MediaMetadata;
import android.media.session.PlaybackState;
import android.media.session.MediaController;
import android.media.session.MediaSessionManager;
import android.os.BatteryManager;
import android.os.Build;
import android.os.Handler;
import android.os.Looper;
import android.provider.Settings;
import android.service.notification.NotificationListenerService;
import android.service.notification.StatusBarNotification;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.view.WindowManager;

import java.util.ArrayDeque;
import java.util.List;
import java.util.Set;

public final class NovaNotificationListenerService extends NotificationListenerService {
    private WindowManager wm;
    private final Handler handler = new Handler(Looper.getMainLooper());
    private android.content.SharedPreferences prefs;
    private NovaSpotView spot;
    private MediaSessionManager mediaManager;
    private MediaController controller;
    private MediaController.Callback controllerCallback;
    private MediaSessionManager.OnActiveSessionsChangedListener sessionsListener;
    private BroadcastReceiver batteryReceiver;
    private Runnable hideRunnable;
    private final ArrayDeque<String> queue = new ArrayDeque<>();
    private boolean draining;

    private final Runnable progressRunnable = new Runnable() {
        @Override public void run() {
            updateMedia();
            if (controller != null && controller.getPlaybackState() != null && controller.getPlaybackState().getState() == PlaybackState.STATE_PLAYING) {
                handler.postDelayed(this, 750);
            }
        }
    };

    @Override public void onListenerConnected() {
        super.onListenerConnected();
        wm = (WindowManager) getSystemService(WINDOW_SERVICE);
        prefs = getSharedPreferences("nova", MODE_PRIVATE);
        registerBattery();
        setupMedia();
        showBattery();
    }

    private void setupMedia() {
        mediaManager = (MediaSessionManager) getSystemService(MEDIA_SESSION_SERVICE);
        final ComponentName component = new ComponentName(this, NovaNotificationListenerService.class);
        sessionsListener = controllers -> selectController(controllers);
        try {
            mediaManager.addOnActiveSessionsChangedListener(sessionsListener, component);
            selectController(mediaManager.getActiveSessions(component));
        } catch (SecurityException ignored) {
        }
    }

    private void selectController(List<MediaController> list) {
        MediaController next = null;
        for (MediaController candidate : list) {
            PlaybackState state = candidate.getPlaybackState();
            if (state != null && state.getState() == PlaybackState.STATE_PLAYING) { next = candidate; break; }
        }
        if (next == null && !list.isEmpty()) next = list.get(0);
        if (sameSession(next, controller)) { updateMedia(); return; }

        if (controller != null && controllerCallback != null) {
            try { controller.unregisterCallback(controllerCallback); } catch (RuntimeException ignored) {}
        }
        controller = next;
        controllerCallback = null;
        handler.removeCallbacks(progressRunnable);

        if (next != null) {
            controllerCallback = new MediaController.Callback() {
                @Override public void onMetadataChanged(MediaMetadata metadata) { updateMedia(); }
                @Override public void onPlaybackStateChanged(PlaybackState state) {
                    updateMedia();
                    handler.removeCallbacks(progressRunnable);
                    if (state != null && state.getState() == PlaybackState.STATE_PLAYING) handler.post(progressRunnable);
                }
            };
            try { next.registerCallback(controllerCallback); } catch (RuntimeException ignored) {}
            PlaybackState state = next.getPlaybackState();
            if (state != null && state.getState() == PlaybackState.STATE_PLAYING) handler.post(progressRunnable);
        }
        updateMedia();
    }

    private boolean sameSession(MediaController a, MediaController b) {
        if (a == null || b == null) return a == b;
        return a.getSessionToken().equals(b.getSessionToken());
    }

    private void updateMedia() {
        if (prefs == null || !prefs.getBoolean("media", true) || controller == null) return;
        MediaMetadata metadata = controller.getMetadata();
        if (metadata == null) return;
        String title = value(metadata.getString(MediaMetadata.METADATA_KEY_TITLE));
        if (title.isEmpty()) return;
        String artist = value(metadata.getString(MediaMetadata.METADATA_KEY_ARTIST));
        long duration = metadata.getLong(MediaMetadata.METADATA_KEY_DURATION);
        if (duration < 0) duration = 0;
        PlaybackState state = controller.getPlaybackState();
        long position = state == null ? 0 : Math.max(0, state.getPosition());
        boolean playing = state != null && state.getState() == PlaybackState.STATE_PLAYING;
        Bitmap art = metadata.getBitmap(MediaMetadata.METADATA_KEY_ART);
        if (art == null) art = metadata.getBitmap(MediaMetadata.METADATA_KEY_ALBUM_ART);
        final String fTitle = title, fArtist = artist;
        final long fDuration = duration, fPosition = position;
        final boolean fPlaying = playing;
        final Bitmap fArt = art;
        handler.post(() -> {
            ensureSpot();
            if (spot != null) {
                spot.setMedia(fTitle, fArtist, fDuration, fPosition, fPlaying, fArt);
                spot.showAnimated();
            }
            if (hideRunnable != null) { handler.removeCallbacks(hideRunnable); hideRunnable = null; }
        });
    }

    @Override public void onNotificationPosted(StatusBarNotification sbn) {
        if (prefs == null || !prefs.getBoolean("notifications", true) || sbn == null || getPackageName().equals(sbn.getPackageName())) return;
        Notification notification = sbn.getNotification();
        if ((notification.flags & Notification.FLAG_ONGOING_EVENT) != 0) return;
        CharSequence titleCs = notification.extras.getCharSequence(Notification.EXTRA_TITLE);
        CharSequence bodyCs = notification.extras.getCharSequence(Notification.EXTRA_TEXT);
        String title = titleCs == null ? "" : titleCs.toString();
        String body = bodyCs == null ? "" : bodyCs.toString();
        if (title.trim().isEmpty() && body.trim().isEmpty()) return;
        String app = "Notification";
        try { app = getPackageManager().getApplicationLabel(getPackageManager().getApplicationInfo(sbn.getPackageName(), 0)).toString(); } catch (Exception ignored) {}
        StringBuilder msg = new StringBuilder("● ").append(app);
        if (!title.trim().isEmpty()) msg.append(": ").append(title);
        if (!body.trim().isEmpty()) msg.append(" — ").append(body);
        enqueue(msg.toString());
    }

    private void registerBattery() {
        if (batteryReceiver != null) return;
        batteryReceiver = new BroadcastReceiver() {
            @Override public void onReceive(Context context, Intent intent) {
                if (prefs == null || !prefs.getBoolean("battery", true)) return;
                int level = intent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1);
                int status = intent.getIntExtra(BatteryManager.EXTRA_STATUS, -1);
                if (level < 0) return;
                if (status == BatteryManager.BATTERY_STATUS_CHARGING || status == BatteryManager.BATTERY_STATUS_FULL) enqueue("⚡ " + level + "%");
                else if (level <= 15) enqueue("🔋 Batterie faible • " + level + "%");
            }
        };
        IntentFilter filter = new IntentFilter(Intent.ACTION_BATTERY_CHANGED);
        if (Build.VERSION.SDK_INT >= 33) registerReceiver(batteryReceiver, filter, Context.RECEIVER_EXPORTED);
        else registerReceiver(batteryReceiver, filter);
    }

    private void showBattery() {
        if (prefs == null || !prefs.getBoolean("battery", true)) return;
        Intent intent = registerReceiver(null, new IntentFilter(Intent.ACTION_BATTERY_CHANGED));
        if (intent == null) return;
        int level = intent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1);
        if (level >= 0) enqueue("🔋 " + level + "%");
    }

    private void enqueue(String text) {
        handler.post(() -> {
            if (queue.size() >= 5) queue.removeFirst();
            queue.addLast(text);
            drain();
        });
    }

    private void drain() {
        if (draining || queue.isEmpty()) return;
        draining = true;
        ensureSpot();
        if (spot == null) { draining = false; queue.clear(); return; }
        spot.setMessage(queue.removeFirst());
        spot.showAnimated();
        if (hideRunnable != null) handler.removeCallbacks(hideRunnable);
        hideRunnable = () -> {
            if (spot != null) spot.hideAnimated();
            draining = false;
            handler.postDelayed(this::drain, 140);
        };
        handler.postDelayed(hideRunnable, 3200);
    }

    private void ensureSpot() {
        if (spot != null || wm == null) return;
        NovaSpotView view = new NovaSpotView(this, this::performMediaAction);
        WindowManager.LayoutParams p = new WindowManager.LayoutParams(
                WindowManager.LayoutParams.WRAP_CONTENT,
                WindowManager.LayoutParams.WRAP_CONTENT,
                WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY,
                WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE,
                android.graphics.PixelFormat.TRANSLUCENT);
        p.gravity = Gravity.TOP | Gravity.CENTER_HORIZONTAL;
        p.y = dp(8);
        try { wm.addView(view, p); spot = view; } catch (RuntimeException ignored) { spot = null; }
    }

    private void performMediaAction(NovaSpotView.Action action) {
        if (controller == null) return;
        try {
            if (action == NovaSpotView.Action.TOGGLE) {
                PlaybackState state = controller.getPlaybackState();
                if (state != null && state.getState() == PlaybackState.STATE_PLAYING) controller.getTransportControls().pause();
                else controller.getTransportControls().play();
            } else if (action == NovaSpotView.Action.NEXT) controller.getTransportControls().skipToNext();
            else controller.getTransportControls().skipToPrevious();
        } catch (RuntimeException ignored) {}
        handler.postDelayed(this::updateMedia, 120);
    }

    @Override public void onDestroy() {
        if (hideRunnable != null) handler.removeCallbacks(hideRunnable);
        handler.removeCallbacks(progressRunnable);
        if (batteryReceiver != null) { try { unregisterReceiver(batteryReceiver); } catch (RuntimeException ignored) {} }
        if (mediaManager != null && sessionsListener != null) { try { mediaManager.removeOnActiveSessionsChangedListener(sessionsListener); } catch (RuntimeException ignored) {} }
        if (controller != null && controllerCallback != null) { try { controller.unregisterCallback(controllerCallback); } catch (RuntimeException ignored) {} }
        if (spot != null && wm != null) { try { wm.removeView(spot); } catch (RuntimeException ignored) {} }
        spot = null;
        super.onDestroy();
    }

    private static String value(String value) { return value == null ? "" : value.trim(); }
    private int dp(int value) { return Math.round(value * getResources().getDisplayMetrics().density); }

    public static boolean isAccessGranted(Context context) {
        try {
            String raw = Settings.Secure.getString(context.getContentResolver(), "enabled_notification_listeners");
            if (raw == null) return false;
            Set<ComponentName> set = new java.util.HashSet<>();
            for (String part : raw.split(":")) {
                ComponentName name = ComponentName.unflattenFromString(part);
                if (name != null) set.add(name);
            }
            return set.contains(new ComponentName(context, NovaNotificationListenerService.class));
        } catch (RuntimeException e) { return false; }
    }

    private static final class NovaSpotView extends View {
        enum Action { TOGGLE, NEXT, PREVIOUS }
        private final float density;
        private final Paint paint = new Paint(Paint.ANTI_ALIAS_FLAG);
        private final Paint bg = new Paint(Paint.ANTI_ALIAS_FLAG);
        private final Paint border = new Paint(Paint.ANTI_ALIAS_FLAG);
        private final ActionHandler handler;
        private boolean media;
        private String message = "", title = "", artist = "";
        private long duration, position;
        private boolean playing;
        private Bitmap art;
        private float downX;
        private long downTime;

        interface ActionHandler { void onAction(Action action); }

        NovaSpotView(Context context, ActionHandler handler) {
            super(context);
            this.handler = handler;
            density = getResources().getDisplayMetrics().density;
            bg.setColor(Color.rgb(16, 18, 21));
            border.setColor(Color.rgb(70, 74, 80));
            border.setStyle(Paint.Style.STROKE);
            border.setStrokeWidth(dpF(1));
            setAlpha(0f); setScaleX(.92f); setScaleY(.92f); setClickable(true);
        }

        void setMessage(String value) { media = false; message = value; requestLayout(); invalidate(); }
        void setMedia(String t, String a, long d, long p, boolean play, Bitmap bitmap) { media = true; title=t; artist=a; duration=d; position=p; playing=play; art=bitmap; requestLayout(); invalidate(); }
        void showAnimated() { setVisibility(VISIBLE); animate().cancel(); animate().alpha(1f).scaleX(1f).scaleY(1f).setDuration(150).start(); }
        void hideAnimated() { animate().alpha(0f).scaleX(.92f).scaleY(.92f).setDuration(130).withEndAction(() -> setVisibility(GONE)).start(); }

        @Override protected void onMeasure(int w, int h) { setMeasuredDimension(dp(media ? 276 : 250), dp(media ? 60 : 42)); }

        @Override protected void onDraw(Canvas canvas) {
            RectF r = new RectF(0, 0, getWidth(), getHeight());
            canvas.drawRoundRect(r, dpF(30), dpF(30), bg);
            canvas.drawRoundRect(r, dpF(30), dpF(30), border);
            if (media) drawMedia(canvas); else drawMessage(canvas);
        }

        private void drawMessage(Canvas c) {
            paint.setColor(Color.WHITE); paint.setTextSize(dpF(13)); paint.setTypeface(android.graphics.Typeface.DEFAULT);
            c.drawText(message.length() > 48 ? message.substring(0,48) : message, dpF(16), dpF(27), paint);
        }

        private void drawMedia(Canvas c) {
            float size = dpF(44), left = dpF(8);
            if (art != null) c.drawBitmap(art, null, new RectF(left, dpF(8), left+size, dpF(52)), paint);
            else { paint.setColor(Color.rgb(44,47,52)); c.drawRoundRect(new RectF(left,dpF(8),left+size,dpF(52)),dpF(12),dpF(12),paint); }
            paint.setColor(Color.WHITE); paint.setTypeface(android.graphics.Typeface.DEFAULT_BOLD); paint.setTextSize(dpF(12.5f));
            c.drawText(shortText(title,25),dpF(60),dpF(24),paint);
            paint.setColor(Color.LTGRAY); paint.setTypeface(android.graphics.Typeface.DEFAULT); paint.setTextSize(dpF(10.5f));
            c.drawText(shortText(artist,26),dpF(60),dpF(39),paint);
            paint.setColor(Color.WHITE); paint.setTextSize(dpF(18));
            c.drawText(playing ? "Ⅱ" : "▶", getWidth()-dpF(48), dpF(31), paint);
            if (duration > 0) {
                float progress = Math.max(0, Math.min(1, (float)position/(float)duration));
                paint.setColor(Color.rgb(92,95,100));
                c.drawRoundRect(new RectF(dpF(60),dpF(49),getWidth()-dpF(58),dpF(52)),dpF(2),dpF(2),paint);
                paint.setColor(Color.WHITE);
                c.drawRoundRect(new RectF(dpF(60),dpF(49),dpF(60)+(getWidth()-dpF(118))*progress,dpF(52)),dpF(2),dpF(2),paint);
            }
        }

        private static String shortText(String s, int max) { return s.length() > max ? s.substring(0,max) : s; }
        @Override public boolean onTouchEvent(MotionEvent e) {
            if (e.getActionMasked() == MotionEvent.ACTION_DOWN) { downX=e.getX(); downTime=System.currentTimeMillis(); return true; }
            if (e.getActionMasked() == MotionEvent.ACTION_UP) {
                if (!media) return true;
                float dx=e.getX()-downX; long elapsed=System.currentTimeMillis()-downTime;
                if (elapsed < 500 && Math.abs(dx) > dpF(45)) handler.onAction(dx > 0 ? Action.NEXT : Action.PREVIOUS);
                else if (elapsed < 500) handler.onAction(Action.TOGGLE);
                return true;
            }
            return true;
        }
        private int dp(float v) { return Math.round(v*density); }
        private float dpF(float v) { return v*density; }
    }
}

File: app/src/main/java/com/rmambulancier/nova/NovaNotificationListenerService.kt
package com.rmambulancier.nova

import android.app.Notification
import android.content.BroadcastReceiver
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.media.MediaMetadata
import android.media.session.MediaController
import android.media.session.MediaSessionManager
import android.os.BatteryManager
import android.os.Handler
import android.os.Looper
import android.provider.Settings
import android.service.notification.NotificationListenerService
import android.service.notification.StatusBarNotification
import android.view.Gravity
import android.view.MotionEvent
import android.view.View
import android.view.WindowManager
import kotlin.math.roundToInt

class NovaNotificationListenerService : NotificationListenerService() {
    private lateinit var windowManager: WindowManager
    private var spot: NovaSpotView? = null
    private val handler = Handler(Looper.getMainLooper())
    private var mediaManager: MediaSessionManager? = null
    private var mediaController: MediaController? = null
    private var mediaCallback: MediaController.Callback? = null
    private var sessionsListener: MediaSessionManager.OnActiveSessionsChangedListener? = null
    private var batteryReceiver: BroadcastReceiver? = null
    private var hideRunnable: Runnable? = null
    private val eventQueue = ArrayDeque<SpotEvent>()
    private var showingEvent = false
    private var connected = false

    private data class SpotEvent(val text: String, val kind: Kind)
    private enum class Kind { NOTIFICATION, BATTERY, CHARGING, LOW_BATTERY }

    private val prefs by lazy { getSharedPreferences("nova", MODE_PRIVATE) }

    override fun onListenerConnected() {
        super.onListenerConnected()
        connected = true
        windowManager = getSystemService(WINDOW_SERVICE) as WindowManager
        registerBattery()
        setupMediaSessions()
        if (prefs.getBoolean("battery", true)) showBatteryNow()
    }

    private fun registerBattery() {
        if (batteryReceiver != null) return
        batteryReceiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context?, intent: Intent?) {
                if (!prefs.getBoolean("battery", true)) return
                val level = intent?.getIntExtra(BatteryManager.EXTRA_LEVEL, -1) ?: -1
                val status = intent?.getIntExtra(BatteryManager.EXTRA_STATUS, -1) ?: -1
                if (level < 0) return
                val charging = status == BatteryManager.BATTERY_STATUS_CHARGING ||
                    status == BatteryManager.BATTERY_STATUS_FULL
                when {
                    charging -> enqueue(SpotEvent("⚡ $level%", Kind.CHARGING))
                    level <= 15 -> enqueue(SpotEvent("🔋 Batterie faible • $level%", Kind.LOW_BATTERY))
                }
            }
        }
        registerReceiver(batteryReceiver, IntentFilter(Intent.ACTION_BATTERY_CHANGED))
    }

    private fun setupMediaSessions() {
        mediaManager = getSystemService(MEDIA_SESSION_SERVICE) as MediaSessionManager
        val component = ComponentName(this, javaClass)
        sessionsListener = MediaSessionManager.OnActiveSessionsChangedListener { controllers ->
            attachBestController(controllers)
        }
        try {
            mediaManager?.addOnActiveSessionsChangedListener(sessionsListener!!, component)
            attachBestController(mediaManager?.getActiveSessions(component).orEmpty())
        } catch (_: SecurityException) {
            // Notification access has not been fully granted yet.
        }
    }

    private fun attachBestController(controllers: List<MediaController>) {
        val candidate = controllers.firstOrNull { it.playbackState?.state == android.media.session.PlaybackState.STATE_PLAYING }
            ?: controllers.firstOrNull()
        if (candidate?.sessionToken == mediaController?.sessionToken) {
            updateMedia()
            return
        }
        mediaCallback?.let { callback -> mediaController?.unregisterCallback(callback) }
        mediaController = candidate
        mediaCallback = candidate?.let {
            object : MediaController.Callback() {
                override fun onMetadataChanged(metadata: MediaMetadata?) = updateMedia()
                override fun onPlaybackStateChanged(state: android.media.session.PlaybackState?) = updateMedia()
            }
        }
        mediaCallback?.let { candidate.registerCallback(it) }
        updateMedia()
    }

    private fun updateMedia() {
        if (!prefs.getBoolean("media", true)) return
        val controller = mediaController ?: return
        val metadata = controller.metadata ?: return
        val title = metadata.getString(MediaMetadata.METADATA_KEY_TITLE).orEmpty()
        if (title.isBlank()) return
        val artist = metadata.getString(MediaMetadata.METADATA_KEY_ARTIST).orEmpty()
        val duration = metadata.getLong(MediaMetadata.METADATA_KEY_DURATION).takeIf { it > 0 } ?: 0L
        val position = controller.playbackState?.position ?: 0L
        val playing = controller.playbackState?.state == android.media.session.PlaybackState.STATE_PLAYING
        val art = metadata.getBitmap(MediaMetadata.METADATA_KEY_ART)
            ?: metadata.getBitmap(MediaMetadata.METADATA_KEY_ALBUM_ART)
        showMedia(title, artist, duration, position, playing, art)
    }

    private fun showMedia(title: String, artist: String, duration: Long, position: Long, playing: Boolean, art: Bitmap?) {
        handler.post {
            ensureSpot()
            spot?.setMedia(title, artist, duration, position, playing, art)
            spot?.animateIn()
            hideRunnable?.let(handler::removeCallbacks)
            hideRunnable = null // Media remains visible while a session is active.
        }
    }

    override fun onNotificationPosted(sbn: StatusBarNotification?) {
        if (sbn == null || sbn.packageName == packageName || !prefs.getBoolean("notifications", true)) return
        val notification = sbn.notification
        if ((notification.flags and Notification.FLAG_ONGOING_EVENT) != 0) return
        val extras = notification.extras
        val title = extras.getCharSequence(Notification.EXTRA_TITLE)?.toString().orEmpty()
        val text = extras.getCharSequence(Notification.EXTRA_TEXT)?.toString().orEmpty()
        if (title.isBlank() && text.isBlank()) return
        val label = runCatching {
            packageManager.getApplicationLabel(packageManager.getApplicationInfo(sbn.packageName, 0)).toString()
        }.getOrDefault("Notification")
        val message = buildString {
            append("● ").append(label)
            if (title.isNotBlank()) append(": ").append(title)
            if (text.isNotBlank()) append(" — ").append(text)
        }
        enqueue(SpotEvent(message, Kind.NOTIFICATION))
    }

    private fun showBatteryNow() {
        val intent = registerReceiver(null, IntentFilter(Intent.ACTION_BATTERY_CHANGED)) ?: return
        val level = intent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1)
        if (level >= 0) enqueue(SpotEvent("🔋 $level%", Kind.BATTERY))
    }

    private fun enqueue(event: SpotEvent) {
        handler.post {
            if (eventQueue.size >= 5) eventQueue.removeFirst()
            eventQueue.addLast(event)
            drainQueue()
        }
    }

    private fun drainQueue() {
        if (showingEvent || eventQueue.isEmpty()) return
        showingEvent = true
        val event = eventQueue.removeFirst()
        ensureSpot()
        spot?.setMessage(event.text)
        spot?.animateIn()
        hideRunnable?.let(handler::removeCallbacks)
        hideRunnable = Runnable {
            spot?.animateOut()
            showingEvent = false
            handler.postDelayed({ drainQueue() }, 120)
        }
        handler.postDelayed(hideRunnable!!, if (event.kind == Kind.LOW_BATTERY) 5000L else 3200L)
    }

    private fun ensureSpot() {
        if (spot != null) return
        val view = NovaSpotView(this) { action -> handleSpotAction(action) }
        spot = view
        val params = WindowManager.LayoutParams(
            WindowManager.LayoutParams.WRAP_CONTENT,
            WindowManager.LayoutParams.WRAP_CONTENT,
            WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY,
            WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE or
                WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS,
            android.graphics.PixelFormat.TRANSLUCENT
        ).apply {
            gravity = Gravity.TOP or Gravity.CENTER_HORIZONTAL
            y = dp(8)
        }
        windowManager.addView(view, params)
    }

    private fun handleSpotAction(action: NovaSpotView.Action) {
        val controller = mediaController ?: return
        runCatching {
            when (action) {
                NovaSpotView.Action.TOGGLE -> {
                    if (controller.playbackState?.state == android.media.session.PlaybackState.STATE_PLAYING) {
                        controller.transportControls.pause()
                    } else controller.transportControls.play()
                }
                NovaSpotView.Action.NEXT -> controller.transportControls.skipToNext()
                NovaSpotView.Action.PREVIOUS -> controller.transportControls.skipToPrevious()
            }
        }
        handler.postDelayed({ updateMedia() }, 120)
    }

    override fun onListenerDisconnected() {
        connected = false
        super.onListenerDisconnected()
    }

    override fun onDestroy() {
        hideRunnable?.let(handler::removeCallbacks)
        batteryReceiver?.let { runCatching { unregisterReceiver(it) } }
        sessionsListener?.let { listener -> runCatching { mediaManager?.removeOnActiveSessionsChangedListener(listener) } }
        mediaCallback?.let { callback -> runCatching { mediaController?.unregisterCallback(callback) } }
        spot?.let { runCatching { windowManager.removeView(it) } }
        spot = null
        super.onDestroy()
    }

    private fun dp(value: Int): Int = (value * resources.displayMetrics.density).roundToInt()

    companion object {
        fun isAccessGranted(context: Context): Boolean =
            runCatching {
                getEnabledListeners(context).contains(ComponentName(context, NovaNotificationListenerService::class.java))
            }.getOrDefault(false)

        private fun getEnabledListeners(context: Context): Set<ComponentName> {
            val flat = Settings.Secure.getString(context.contentResolver, "enabled_notification_listeners") ?: return emptySet()
            return flat.split(":").mapNotNull { ComponentName.unflattenFromString(it) }.toSet()
        }
    }

    private class NovaSpotView(
        context: Context,
        private val actionHandler: (Action) -> Unit
    ) : View(context) {
        enum class Action { TOGGLE, NEXT, PREVIOUS }

        private val density = resources.displayMetrics.density
        private val paint = Paint(Paint.ANTI_ALIAS_FLAG)
        private val bg = Paint(Paint.ANTI_ALIAS_FLAG).apply { color = Color.rgb(18, 20, 23) }
        private val stroke = Paint(Paint.ANTI_ALIAS_FLAG).apply { color = Color.rgb(72, 76, 82); style = Paint.Style.STROKE; strokeWidth = dp(1f) }
        private var message: String = ""
        private var title: String = ""
        private var artist: String = ""
        private var duration: Long = 0L
        private var position: Long = 0L
        private var playing: Boolean = false
        private var artwork: Bitmap? = null
        private var mediaMode = false
        private var downX = 0f
        private var downY = 0f
        private var downTime = 0L

        init {
            isClickable = true
            alpha = 0f
            scaleX = .9f
            scaleY = .9f
        }

        fun setMessage(value: String) {
            mediaMode = false
            message = value
            requestLayout(); invalidate()
        }

        fun setMedia(t: String, a: String, d: Long, p: Long, isPlaying: Boolean, art: Bitmap?) {
            mediaMode = true
            title = t; artist = a; duration = d; position = p; playing = isPlaying; artwork = art
            requestLayout(); invalidate()
        }

        fun animateIn() {
            visibility = VISIBLE
            animate().cancel()
            animate().alpha(1f).scaleX(1f).scaleY(1f).setDuration(160).start()
        }

        fun animateOut() {
            animate().alpha(0f).scaleX(.9f).scaleY(.9f).setDuration(140).withEndAction { visibility = GONE }.start()
        }

        override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
            val width = if (mediaMode) dp(270f) else dp(250f)
            setMeasuredDimension(width, dp(if (mediaMode) 58f else 42f))
        }

        override fun onDraw(canvas: Canvas) {
            super.onDraw(canvas)
            val r = RectF(0f, 0f, width.toFloat(), height.toFloat())
            canvas.drawRoundRect(r, dp(29f), dp(29f), bg)
            canvas.drawRoundRect(r, dp(29f), dp(29f), stroke)
            if (mediaMode) drawMedia(canvas) else drawMessage(canvas)
        }

        private fun drawMessage(canvas: Canvas) {
            paint.color = Color.WHITE
            paint.textSize = dp(13f)
            paint.typeface = android.graphics.Typeface.create("sans", android.graphics.Typeface.NORMAL)
            canvas.drawText(message.take(48), dp(16f), dp(27f), paint)
        }

        private fun drawMedia(canvas: Canvas) {
            val artSize = dp(42f)
            val left = dp(8f)
            artwork?.let {
                val dst = RectF(left, dp(8f), left + artSize, dp(50f))
                canvas.save()
                canvas.clipRect(dst)
                canvas.drawBitmap(it, null, dst, paint)
                canvas.restore()
            } ?: run {
                paint.color = Color.rgb(45, 48, 53)
                canvas.drawRoundRect(RectF(left, dp(8f), left + artSize, dp(50f)), dp(12f), dp(12f), paint)
            }

            paint.color = Color.WHITE
            paint.textSize = dp(12.5f)
            paint.typeface = android.graphics.Typeface.DEFAULT_BOLD
            canvas.drawText(title.take(25), dp(58f), dp(23f), paint)
            paint.typeface = android.graphics.Typeface.DEFAULT
            paint.color = Color.LTGRAY
            paint.textSize = dp(10.5f)
            canvas.drawText(artist.take(27), dp(58f), dp(38f), paint)

            paint.color = Color.WHITE
            paint.textSize = dp(18f)
            val icon = if (playing) "Ⅱ" else "▶"
            canvas.drawText(icon, width - dp(48f), dp(30f), paint)

            if (duration > 0) {
                val progress = (position.toFloat() / duration.toFloat()).coerceIn(0f, 1f)
                paint.color = Color.rgb(105, 105, 110)
                canvas.drawRoundRect(RectF(dp(58f), dp(47f), width - dp(58f), dp(50f)), dp(2f), dp(2f), paint)
                paint.color = Color.WHITE
                canvas.drawRoundRect(RectF(dp(58f), dp(47f), dp(58f) + (width - dp(116f)) * progress, dp(50f)), dp(2f), dp(2f), paint)
            }
        }

        override fun onTouchEvent(event: MotionEvent): Boolean {
            when (event.actionMasked) {
                MotionEvent.ACTION_DOWN -> {
                    downX = event.x; downY = event.y; downTime = System.currentTimeMillis(); return true
                }
                MotionEvent.ACTION_UP -> {
                    if (!mediaMode) return true
                    val dx = event.x - downX
                    val elapsed = System.currentTimeMillis() - downTime
                    if (elapsed < 500 && kotlin.math.abs(dx) > dp(45f)) {
                        actionHandler(if (dx > 0) Action.NEXT else Action.PREVIOUS)
                    } else if (elapsed < 500) {
                        actionHandler(Action.TOGGLE)
                    }
                    return true
                }
            }
            return true
        }

        private fun dp(value: Float): Float = value * density
    }
}
